### ISSUE TITLE
Dependencies handling in cmake config

### ISSUE THREAD

**Issue by mitsch** [davisking/dlib#1594]
When installing dlib, I see that the exported cmake config has dependencies with hardcoded paths imprinted. This causes any other environment setting to keep the dependencies at the same place.

In my case, I have  a docker where I install dlib with cudnn at /usr/libx86_64-linux-gnu.
On the system I want my dlib to deploy to, I have cudnn at /opt/cudnn-7.

Usually cmake allows to abstract from this path-hardcoded dependency by letting dlibConfig.cmake find cudnn a second time (when find_package(dlib ...) is called).

Is there a reason why this flexibility is not implemented? (besides the comfort of not explicitely declaring the dependencies a second time)

**Comment by xsacha** [comment]
## DLib looks here                                          SET(CMAKE_PREFIX_PATH "${CMAKE_PREFIX_PATH} ${PREBUILT}/cuda/")

**Comment by xsacha** [comment]
CUDNN is not part of or distributed with dlib, so the CUDNN would need to be found again on the system where it is being imported, right?

**Comment by mitsch** [comment]
Exactly.

I understand that dlib so far assumes that it will be always at the path where cmake found it at configuration time of dlib compilation/installation. In my case that is `/usr/lib/x86_64-linux-gnu/libcudnn.so`. Shipping all libraries/headers/cmakeConfigs to another system, cudnn needs to be found at a different place, but dlib.cmake is assuming thtat libcudnn.so is at /usr/lib/x86_64-linux-gnu/libcudnn.so. At least it is hardcoded in my ${CMAKE_INSTALL_PREFIX}/lib/cmake/dlib/dlib.cmake.

It most likely is a small issue for most people, since they install and use dlib at the same system. I wanted to package it and came across that problem.

Is it an issue you are aware of or even have decided to do it that way? (out of curiosity)

**Comment by xsacha** [comment]
Yeah so the problem is that these are full paths rather than just the name of the library file so you can resolve it on the other end.

I suppose the main issue is there is no FindCUDNN or FindMKL.

**Comment by davisking** [comment]
I think this is generally how cmake installs work. If you want to propose a change then submit a PR. Note that it must work on all platforms and not break any existing use cases.

**Comment by mitsch** [comment]
I think I know now why CMake works that way. Using the cmake2 way of setting variables (in the Find* files causes the hardcoded paths. I'm gonna try in the new year to submit a PR where the Find* files are defining target and properties. This I reckon should help to erase the hardcoded paths from the Config*.
